File: solutions/python/proverb/1/proverb.py
def proverb(*args, qualifier=None):
    if not args:
        return []
    
    res = []
    for i in range(len(args) - 1):
        res.append(f"For want of a {args[i]} the {args[i+1]} was lost.")
        
    final_item = f"{qualifier} {args[0]}" if qualifier else args[0]
    res.append(f"And all for the want of a {final_item}.")
    
    return res

File: solutions/ruby/proverb/1/proverb.rb
class Proverb
  def initialize(*args, qualifier: nil)
    @list = args
    @qualifier = qualifier
  end

  def to_s
    return "" if @list.empty?
    
    result = ""
    (0...@list.length - 1).each do |i|
      result += "For want of a #{@list[i]} the #{@list[i+1]} was lost.\n"
    end
    
    last = @qualifier ? "#{@qualifier} #{@list[0]}" : @list[0]
    result += "And all for the want of a #{last}."
    result
  end
end
